File: app/src/main/java/com/ballastlane/android/texasam/data/managers/remote/base/RestClient.java
package com.ballastlane.android.texasam.data.managers.remote.base;

import com.ballastlane.android.baselibrary.model.RestClientManager;

/**
 * @author Daniela Perez danielaperez@kogimobile.com on 4/24/18.
 */
public class RestClient {

    private static RestClient instance = null;
    private static RestClientManager restClientManager;

//    @NonNull
//    public static RestClient getInstance() {
//        if (instance == null) {
//            instance = new RestClient();
//            ServiceConfiguration<RestManagerPublicService, ApiServiceInterface> publicServiceConfiguration;
//            publicServiceConfiguration = new ServiceConfiguration<>(
//                    "http://172.23.10.45:1337/v1/",
//                    new ApiServiceFactory<>(RestManagerPublicService.class),
//                    ApiServiceInterface.class);
//
//            ServiceConfiguration<RestManagerPrivateService, ApiServiceInterface> privateServiceConfiguration =
//                    new ServiceConfiguration<>(
//                            "http://172.23.10.45:1337/v1/",
//                            new ApiServiceFactory<>(RestManagerPrivateService.class),
//                            ApiServiceInterface.class);
//
//            privateServiceConfiguration.setRequestInterceptor(new Interceptor() {
//                @Override
//                public Response intercept(Chain chain) throws IOException {
//                    Request request = chain.request();
//                    Request.Builder builder = (new Request.Builder()).url(request.url()).method(request.method(), request.body());
//                    String accessToken = PreferencesUtils.getUserToken();
//
//                    if (!accessToken.isEmpty()) {
//                        Header currentHeader = new Header("Authorization", String.format("%s", accessToken));
//                        builder.addHeader(currentHeader.name.utf8(), currentHeader.value.utf8());
//                    }
//                    return chain.proceed(builder.build());
//                }
//            });
//            restClientManager = new RestClientManager();
//            restClientManager.addServiceConfiguration(publicServiceConfiguration);
//            restClientManager.addServiceConfiguration(privateServiceConfiguration);
//        }
//        return instance;
//    }

    public static RestClientManager getRestClientManager() {
        return restClientManager;
    }
}

File: sample/src/main/java/com/ballastlane/ballastlanebaseapp/data/managers/remote/RestClient.kt
package com.ballastlane.ballastlanebaseapp.data.managers.remote

import com.kogimobile.baselibrary.sample.R
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import okhttp3.internal.http2.Header
import java.io.IOException

/**
 * Created by Mariangela Salcedo (mariangelasalcedo@ballastlane.com) on 4/2/18.
 * Copyright (c) 2018 Ballast Lane Applications LLC. All rights reserved.
 */

class RestClient {

    companion object {
        private var instance: RestClient? = null
        var restClientManager: RestClientManager? = null
    }

    fun getInstance(): RestClient {
        if (instance == null) {
            instance = RestClient()
            val publicServiceConfiguration = ServiceConfiguration(
                    App.getGlobalContext().getString(R.string.base_url),
                    ApiServiceFactory(RestManagerPublicService::class.java),
                    ApiServiceInterface::class.java)

            val privateServiceConfiguration = ServiceConfiguration(
                    App.getGlobalContext().getString(R.string.base_url),
                    ApiServiceFactory(RestManagerPrivateService::class.java),
                    ApiServiceInterface::class.java)

            privateServiceConfiguration.requestInterceptor = Interceptor { chain ->
                val request = chain.request()
                val builder = Request.Builder().url(request.url()).method(request.method(), request.body())
                val accessToken = "" // set access token
                val tokenType = "" // set token type

                if (tokenType.isNotEmpty() && accessToken.isNotEmpty()) {
                    val currentHeader = Header("Authorization", String.format("%s %s", tokenType, accessToken))
                    builder.addHeader(currentHeader.name.utf8(), currentHeader.value.utf8())
                }
                chain.proceed(builder.build())
            }

            restClientManager = RestClientManager()
            restClientManager!!.addServiceConfiguration(publicServiceConfiguration)
            restClientManager!!.addServiceConfiguration(privateServiceConfiguration)
        }
        return instance as RestClient

    }
}
